File: src/medium/Q0011.cpp
// 11. Container With Most Water
// https://leetcode.com/problems/container-with-most-water/

#include <algorithm>
#include <vector>

using namespace std;

// Time: O(n), Space: O(1), where n is the size of `height`.
auto max_area(vector<int> &height) -> int {
  if (height.size() < 2) {
    return -1;
  }
  int result = 0;
  size_t left = 0, right = height.size() - 1;
  while (left < right) {
    const int x = right - left;
    const int y = min(height[left], height[right]);
    result = max(result, x * y);
    if (height[left] < height[right]) {
      ++left;
    } else {
      --right;
    }
  }
  return result;
}

File: src/main/java/medium/Q0011.java
package medium;

/**
 * 11. Container With Most Water
 * https://leetcode.com/problems/container-with-most-water/
 */
class Q0011 {

  // O(n) time, O(1) space, where n is the length of `height`
  int maxArea(int[] height) {
    int l = 0, r = height.length - 1, maxArea = 0;
    while (l < r) {
      int area = Math.min(height[l], height[r]) * (r - l);
      maxArea = Math.max(maxArea, area);
      if (height[l] < height[r]) {
        l += 1;
      } else {
        r -= 1;
      }
    }
    return maxArea;
  }
}
